tests/nested/core/core20-fault-inject-on-refresh/task.yaml

```diff
@@ -0,0 +1,175 @@
+    Environment=SNAPD_FAULT_INJECT=$TAG:$FAULT
+    EOF
+
+    echo "Wait for the system to be seeded first"
+    remote.exec "sudo snap wait system seed.loaded"
+
+    remote.push fault-inject.conf
+    remote.exec "sudo mkdir -p /etc/systemd/system/snapd.service.d"
+    remote.exec "sudo cp -v fault-inject.conf /etc/systemd/system/snapd.service.d/"
+    remote.exec "sudo systemctl daemon-reload"
+    remote.exec "sudo systemctl restart snapd.service"
+
+execute: |
+    VERSION="$(tests.nested show version)"
+    SNAP=
+    CHANNEL=edge
+    REBOOT=false
+    SECOND_REBOOT=false
+
+    if [ "$FAULT" = reboot ]; then
+        REBOOT=true
+    fi
+
+    case "$SPREAD_VARIANT" in
+        base_*)
+            SNAP="core${VERSION}"
+            if [ "$FAULT" = reboot ]; then
+                SECOND_REBOOT=true
+            else
+                REBOOT=true
+            fi
+            ;;    
+        gadget_*)
+            SNAP=pc
+            CHANNEL="$VERSION/$CHANNEL"
+            ;;
+        kernel_*)
+            SNAP=pc-kernel
+            CHANNEL="$VERSION/beta"
+            if [ "$FAULT" = reboot ]; then
+                SECOND_REBOOT=true
+            else
+                REBOOT=true
+            fi
+            ;;
+        snapd_*)
+            SNAP=snapd
+            ;;
+        *)
+            echo "scenario no supported: $SPREAD_VARIANT"
+            exit 1
+            ;;
+    esac
+        
+    # Get the initial snap revision
+    INITIAL_REV="$(remote.exec snap list | grep -E '^snapd' | awk ' {print $3} ')"
+
+    # Refresh and reboot before the process is completed
+    boot_id="$(tests.nested boot-id)"
+    change_id="$(remote.exec "sudo snap refresh --amend --channel $CHANNEL --no-wait $SNAP")"
+
+    if [ "$REBOOT" = true ]; then
+        remote.wait-for reboot "$boot_id"
+        boot_id="$(tests.nested boot-id)"
+    fi
+
+    if [ "$SECOND_REBOOT" = true ]; then
+        remote.wait-for reboot "$boot_id"
+    fi
+
+    echo "And snap refresh is completed"
+    remote.exec "snap watch $change_id" || true
+    # shellcheck disable=SC2016
+    retry --wait 1 -n 60 --env "CHANGE_ID=$change_id" --env "SNAP=$SNAP" sh -c 'remote.exec snap changes | MATCH "${CHANGE_ID} .* (Done|Error) .*"'
+
+    # Check if the fault was injected
+    remote.exec "test -e /var/lib/snapd/faults/$TAG:$FAULT"
+
+    if [ "$FAULT" = panic ]; then
+        echo "Ensure the panic was injected and triggered stamp file is present"
+        remote.exec "sudo journalctl -u snapd | grep -A 3 panic" > output
+        MATCH "$TAG":$FAULT < output
+        MATCH "osutil.injectFault" < output
+    fi
+
+    FINAL_STATUS="$(remote.exec snap changes | grep -E "^$change_id .*" | awk ' {print $2} ')"
+    CURRENT_REV="$(remote.exec snap list | grep -E '^snapd' | awk ' {print $3} ')"
+
+    if [ "$FINAL_STATUS" = Error ]; then
+        echo "Ensure the change has steps with Error"
+        remote.exec "snap change $change_id" | MATCH "^Error .*"
```

Comment:
hm wonder if `snap watch $change_id` would return the right status and show the error message?

Reply:
snap watch just shows an error message in case of error, not the real state, perhaps checking the message could be a bit dangerous